src/main/kotlin/org/pkl/lsp/completion/UnqualifiedAccessCompletionProvider.kt

```diff
@@ -239,48 +256,391 @@ class UnqualifiedAccessCompletionProvider(private val project: Project) : Comple
     return isClassOrTypeAlias(type)
   }
 
+  private fun inClassBody(node: PklNode): Boolean =
+    node is PklClassBody || (node is PklError && node.parent is PklClassBody)
+
+  private fun inObjectBody(node: PklNode): Boolean =
+    node is PklObjectBody || node.parent is PklObjectElement
+
+  /*
+  This is a heuristic to detect if this node is a top-level definition.
+  We can't just rely upon `node is PklModule` because there are some
+  cases where it doesn't work: `function f() = <caret>`
+   */
+  private fun inTopLevelModule(node: PklNode, col: Int): Boolean =
+    (node is PklModule && col == 1) || (node is PklError && node.parent is PklModule)
+
+  private fun addDefinitionCompletions(
+    node: PklNode,
+    thisType: Type,
+    alreadyDefinedProperties: Set<String>,
+    base: PklBaseModule,
+    collector: MutableList<CompletionItem>,
+    context: PklProject?,
+  ): Boolean {
+
+    return when {
+      thisType == Type.Unknown -> false
+      thisType is Type.Union ->
+        addDefinitionCompletions(
+          node,
+          thisType.leftType,
+          alreadyDefinedProperties,
+          base,
+          collector,
+          context,
+        ) &&
+          addDefinitionCompletions(
+            node,
+            thisType.rightType,
+            alreadyDefinedProperties,
+            base,
+            collector,
+            context,
+          )
+      thisType.isSubtypeOf(base.typedType, base, context) -> {
+        val enclosingDef =
+          node.parentOfTypes(
+            PklModule::class,
+            PklClass::class,
+            // stop classes
+            PklExpr::class,
+            PklObjectBody::class,
+          )
+        val isClassDef =
+          when (enclosingDef) {
+            is PklModule -> !enclosingDef.isAmend
+            is PklClass -> true
+            else -> false
+          }
+        addTypedCompletions(
+          alreadyDefinedProperties,
+          isClassDef,
+          thisType,
+          base,
+          collector,
+          context,
+        )
+        collector.addAll(DEFINITION_LEVEL_KEYWORD_ELEMENTS)
+
+        when (node) {
+          is PklExpr -> {}
+          is PklObjectBody -> {}
+          else -> {
+            when (enclosingDef) {
+              is PklModule -> {
+                collector.addAll(MODULE_LEVEL_KEYWORD_ELEMENTS)
+              }
+              is PklClass -> {
+                collector.addAll(CLASS_LEVEL_KEYWORD_ELEMENTS)
+              }
+              else -> {}
+            }
+          }
+        }
```

Comment:
Feels like a case where `if` might actually be quicker _and_ cleaner... Not blocking, just saying.